core/trino-main/src/main/java/io/trino/operator/join/SortedPositionLinks.java

```diff
@@ -109,31 +110,47 @@ public Factory build()
             ArrayPositionLinks.FactoryBuilder arrayPositionLinksFactoryBuilder = ArrayPositionLinks.builder(size);
             int[][] sortedPositionLinks = new int[size][];
 
-            for (Int2ObjectMap.Entry<IntArrayList> entry : positionLinks.int2ObjectEntrySet()) {
+            Iterator<Int2ObjectMap.Entry<IntArrayList>> iterator = positionLinks.int2ObjectEntrySet().fastIterator();
+            while (iterator.hasNext()) {
+                Int2ObjectOpenHashMap.Entry<IntArrayList> entry = iterator.next();
                 int key = entry.getIntKey();
                 IntArrayList positions = entry.getValue();
-                positions.sort(comparator);
 
-                sortedPositionLinks[key] = new int[positions.size()];
-                for (int i = 0; i < positions.size(); i++) {
-                    sortedPositionLinks[key][i] = positions.getInt(i);
-                }
-
-                // ArrayPositionsLinks.Builder::link builds position links from
-                // tail to head, so we must add them in descending order to have
-                // smallest element as a head
-                for (int i = positions.size() - 2; i >= 0; i--) {
-                    arrayPositionLinksFactoryBuilder.link(positions.getInt(i), positions.getInt(i + 1));
-                }
-
-                // add link from starting position to position links chain
-                if (!positions.isEmpty()) {
-                    arrayPositionLinksFactoryBuilder.link(key, positions.getInt(0));
+                int[] sortedPositions = positions.toIntArray();
+                sortedPositionLinks[key] = sortedPositions;
+
+                if (sortedPositions.length > 0) {
+                    // Single element arrays need not be sorted and have no internal pairs to link
+                    if (sortedPositions.length > 1) {
+                        // Use the positions elements array for the merge sort temporary work buffer to avoid an extra redundant
+                        // copy. This works because we know that initially it has the same values as the array being sorted
+                        IntArrays.mergeSort(sortedPositions, 0, sortedPositions.length, comparator, positions.elements());
+                        // ArrayPositionsLinks.Builder::link builds position links from
+                        // tail to head, so we must add them in descending order to have
+                        // smallest element as a head
+                        for (int i = sortedPositions.length - 2; i >= 0; i--) {
```

Comment:
`link` is just
```
        @Override
        public int link(int left, int right)
        {
            size++;
            positionLinks[left] = right;
            return left;
        }
```
why can't we do:
```
for (int i = 0; i < sortedPositions.length - 1; i++) {
```
?

Reply:
Yeah, I have no idea. I chose to respect the comment above and no think too hard about it- but it does seem like there should be no reason to do this... maybe it's an artifact of a previous `ArrayPositionLinks` implementation?